### PR TITLE
`new` TwintSDK integration

### PR DESCRIPTION
**Pull request by neelSharma12**
<new>

- Support for Twint SDK integration through the `TwintComponent` module.

</new>

### CODE REVIEW

File: AdyenTwint/TwintComponent.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+public final class TwintComponent {
+
+    private lazy var formViewController: FormViewController = {
+
+        return formViewController
+    }()
+
+    public var requiresModalPresentation: Bool = true
+
+    /// Initializes th
```
Comment (erenbesel) [inline AdyenTwint/TwintComponent.swift]: Is this the whole file for the component? Seems empty :D
also missing documentation at the beginning.

File: AdyenTwint/AdyenTwint.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
+//
+// Copyright (c) 2023 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+#import <Foundation/Foundation.h>
+
+//! Project version number for AdyenTwint.
+FOUNDATION_EXPORT double AdyenTwintVersionNumber;
```
Comment (erenbesel) [inline AdyenTwint/AdyenTwint.h:10]: Hmm, why do we need these values for Twint?

File: AdyenTwint/TwintComponent.swift (hunk 1 above)
Comment (neelSharma12) [reply AdyenTwint/TwintComponent.swift]: This is just a placeholder file. I am going to make changes to this in upcoming PR.

File: AdyenTwint/AdyenTwint.h (hunk 2 above)
Comment (neelSharma12) [reply AdyenTwint/AdyenTwint.h:10]: These files came automatically when I created the new target for Twint.

File: AdyenTwint/TwintComponent.swift [hunk 3]
```diff
@@ -0,0 +1,18 @@
+//
+// Copyright (c) 2023 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+@_spi(AdyenInternal) import Adyen
+import TwintSDK
+
+public final class TwintComponent {
```
Comment (goergisn) [inline AdyenTwint/TwintComponent.swift]: Is this missing the actual implementation?

Comment (neelSharma12) [reply AdyenTwint/TwintComponent.swift]: yes, this will contain the code. This PR is only for integration. There is no flow implementation in this.

File: Adyen/Assets/Generated/LocalizationKey.swift [hunk 4]
```diff
@@ -375,6 +374,8 @@ public struct LocalizationKey {
     public static let cashAppPayTitle = LocalizationKey(key: "adyen.cashAppPay.title")
     /// Cashtag
     public static let cashAppPayCashtag = LocalizationKey(key: "adyen.cashAppPay.cashtag")
+    /// No or an outdated version of TWINT is installed on this device. Please update or install the TWINT app.
```
Comment (erenbesel) [inline Adyen/Assets/Generated/LocalizationKey.swift]: >No or an outdated

Feels very weird to read :D

File: AdyenTwint/AdyenTwint.docc/AdyenTwint.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+<string>twint-issuer41</string>
+<string>twint-issuer42</string>
+<string>twint-issuer43</string>
+<string>twint-issuer44</string>
+<string>twint-issuer45</string>
+<string>twint-issuer46</string>
+<string>twint-issuer47</string>
+<string>twint-issuer48</string>
+<string>twint-issuer49</string>
+<string>twint-issuer50</string>
```
Comment (atmamont) [inline AdyenTwint/AdyenTwint.docc/AdyenTwint.md:112]: This could be confusing in the end to any iOS developer who is aware of the limitation of 50 entries. Let's think of some advices we can provide merchant teams with? Like any idea if we can have a mapping of url scheme and a specific app provider? Like which of them to sacrifice?

File: Demo/UIKit/Info.plist [hunk 6]
```diff
@@ -73,6 +73,56 @@
...
+		<string>twint-issuer9</string>
+		<string>twint-issuer8</string>
+		<string>twint-issuer7</string>
+		<string>twint-issuer6</string>
+		<string>twint-issuer5</string>
+		<string>twint-issuer4</string>
+		<string>twint-issuer3</string>
+		<string>twint-issuer2</string>
+		<string>twint-issuer1</string>
 		<string>weixinULAPI</string>
```
Comment (atmamont) [inline Demo/UIKit/Info.plist:126]: Did you test `weixinULAPI` and the following one? They should not be handled now I think according to Apple.

**Review by atmamont — DISMISSED** [review]
Great job Neelam and Alex!
It was really helpful to have intro session, since then this became so much more clear despite the PR size.

File: Demo/UIKit/Info.plist (hunk 6 above)
Comment (neelSharma12) [reply Demo/UIKit/Info.plist:126]: `weixinULAPI` and the following one are not related to Twint.

File: AdyenTwint/AdyenTwint.docc/AdyenTwint.md (hunk 5 above)
Comment (neelSharma12) [reply AdyenTwint/AdyenTwint.docc/AdyenTwint.md:112]: I will come back to you on this.

File: AdyenActions/AdyenActionComponent.swift
Comment (descorp) [inline AdyenActions/AdyenActionComponent.swift:1]: I think it is time to move `AdyenActionComponent.Confioguration` in a separated file 😅

File: AdyenActions/AdyenActionComponent.swift [hunk 7]
```diff
@@ -65,18 +65,55 @@ public final class AdyenActionComponent: ActionComponent, ActionHandlingComponen
...
+                if !Self.isCallbackSchemeValid(callbackAppScheme) {
+                    AdyenAssertion.assertionFailure(message: "Format of provided callbackAppScheme '\(callbackAppScheme)' is incorrect.")
+                }
+                
+                self.callbackAppScheme = callbackAppScheme
+            }
+            
+            /// Validating whether or not the provided `callbackAppScheme` only contains a scheme
+            private static func isCallbackSchemeValid(_ callbackAppScheme: String) -> Bool {
+                if let url = URL(string: callbackAppScheme), url.scheme != nil {
```
Comment (descorp) [inline AdyenActions/AdyenActionComponent.swift:91]: We can also use `UIApplication.shared.canOpenURL(url)` here.
I would suggest appending `AnyAppLauncher` protocol with `canOpen` method - this will improve testability of this code.

File: Demo/UIKit/Info.plist (hunk 6 above)
Comment (descorp) [reply Demo/UIKit/Info.plist:126]: `weixinULAPI` is for WeChat

File: AdyenTwint/TwintDetails.swift [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+                subType: String) {
+        self.type = type.type
+        self.subType = subType
+    }
+
+    // MARK: - Private
+
+    private enum CodingKeys: String, CodingKey {
+        case type
+        case subType = "subtype"
```
Comment (descorp) [inline AdyenTwint/TwintDetails.swift:35]: `checkoutAttemptId` is not part of **CodingKeys**, will/should it be added into JSON ?

File: Demo/UIKit/Info.plist (hunk 6 above)
Comment (atmamont) [reply Demo/UIKit/Info.plist:126]: We now have 52 entries in the plist file and Apple supports 50. What would happen if we try paying via WeChat?

Comment (atmamont) [reply Demo/UIKit/Info.plist:126]: From [the docs](https://developer.apple.com/documentation/uikit/uiapplication/1622952-canopenurl):
> Apps linked on or after iOS 15 are limited to a maximum of 50 entries in the LSApplicationQueriesSchemes key

Comment (descorp) [reply Demo/UIKit/Info.plist:126]: IMHO - this doesn't sound like Adyen problem 😅

File: AdyenActions/Components/SDK/TwintSDKActionComponent.swift [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+                    self.delegate?.didFail(
+                        with: ComponentError.cancelled,
+                        from: self
+                    )
+                }
+            )
+            
+            if let viewController = appChooserViewController, let delegate = presentationDelegate {
+                present(viewController, presentationDelegate: delegate)
+            }
```
Comment (descorp) [inline AdyenActions/Components/SDK/TwintSDKActionComponent.swift:164]: Maybe change this to "guard" and log warning on "else"? Otherwise nothing will happen and merchant will have no clue why..

File: Adyen/Core/Components/Base/PresentableComponentWrapper.swift [hunk 10]
```diff
@@ -24,7 +24,7 @@ public final class PresentableComponentWrapper: PresentableComponent,
     /// The wrapped component.
     public let component: Component
     
-    public let requiresModalPresentation: Bool = true
+    public var requiresModalPresentation: Bool = true
```
Comment (descorp) [inline Adyen/Core/Components/Base/PresentableComponentWrapper.swift:27]: Why this change is needed?

Comment (neelSharma12) [reply Adyen/Core/Components/Base/PresentableComponentWrapper.swift:27]: To hide the  topbar/toolbar from the Twint app chooser VC UI.

File: AdyenActions/Components/SDK/TwintSDKActionComponent.swift (hunk 9 above)
Comment (goergisn) [reply AdyenActions/Components/SDK/TwintSDKActionComponent.swift:164]: The `appChooserViewController` is only optional because of Objective-C (It is never™️ nil) and the presentationDelegate is already checked/asserted on before 👌

File: AdyenActions/AdyenActionComponent.swift (hunk 7 above)
Comment (goergisn) [reply AdyenActions/AdyenActionComponent.swift:91]: Afaik `UIApplication.shared.canOpenURL(url)` only works as expected when you add the url you want to check to the `LSApplicationQueriesSchemes` of your plist. But let me double check (happy to not reinvent the wheel 😅)

File: AdyenTwint/TwintDetails.swift (hunk 8 above)
Comment (neelSharma12) [reply AdyenTwint/TwintDetails.swift:35]: As discussed this is not needed.

File: AdyenActions/AdyenActionComponent.swift
Comment (neelSharma12) [reply AdyenActions/AdyenActionComponent.swift:1]: yes, I will discuss this in our next standup meeting and then take action :)

File: Adyen.podspec [hunk 11]
```diff
@@ -45,6 +45,18 @@ Pod::Spec.new do |s|
...
   end
 
+  s.subspec 'AdyenTwint' do |plugin|
+    plugin.source_files = 'AdyenTwint/**/*.swift'
+    plugin.dependency 'Adyen/Core'
+    plugin.dependency 'Adyen/Actions'
+    plugin.dependency 'Adyen/TwintSDK'
+  end
+
+  s.subspec 'TwintSDK' do |plugin|
```
Comment (erenbesel) [inline Adyen.podspec]: Can this subspec be put into AdyenTwint subspec to prevent this from being podded directly?

File: Adyen/Core/Payment Methods/Abstract/AnyPaymentMethod.swift [hunk 12]
```diff
@@ -39,7 +39,10 @@ internal enum AnyPaymentMethod: Codable {
     case onlineBanking(OnlineBankingPaymentMethod)
     case upi(UPIPaymentMethod)
     case cashAppPay(CashAppPayPaymentMethod)
-    
+    #if canImport(TwintSDK)
```
Comment (erenbesel) [inline Adyen/Core/Payment Methods/Abstract/AnyPaymentMethod.swift]: We potentially don't need canImport checks here, similar to Cashapp. Otherwise there will be many more canImport checks that can be avoided.

File: Package.swift [hunk 13]
```diff
@@ -180,6 +184,19 @@ let package = Package(
...
+        .target(
+            name: "AdyenTwint",
+            dependencies: [
+                .target(name: "Adyen"),
+                .target(name: "AdyenActions"),
+                .target(name: "TwintSDK")
+            ],
+            path: "AdyenTwint"
+        ),
+        .binaryTarget(
```
Comment (erenbesel) [inline Package.swift:191]: Should we add both static and dynamic options?

File: AdyenActions/Actions/SDKAction.swift [hunk 14]
```diff
@@ -12,13 +12,24 @@ public enum SDKAction: Decodable {
     /// Indicates a WeChat Pay SDK action.
     case weChatPay(WeChatPaySDKAction)
     
+    #if canImport(TwintSDK)
```
Comment (erenbesel) [inline AdyenActions/Actions/SDKAction.swift]: same here

File: AdyenActions/Actions/SDKAction.swift [hunk 15]
```diff
@@ -28,5 +39,8 @@ public enum SDKAction: Decodable {
     
     private enum SDKType: String, Decodable {
         case weChatPay = "wechatpaySDK"
+        #if canImport(TwintSDK)
```
Comment (erenbesel) [inline AdyenActions/Actions/SDKAction.swift]: same here

File: AdyenActions/AdyenActionComponent.swift [hunk 16]
```diff
@@ -176,6 +213,12 @@ public final class AdyenActionComponent: ActionComponent, ActionHandlingComponen
         switch sdkAction {
         case let .weChatPay(weChatPaySDKAction):
             handle(weChatPaySDKAction)
+        // swiftlint:disable switch_case_alignment
+        #if canImport(TwintSDK)
```
Comment (erenbesel) [inline AdyenActions/AdyenActionComponent.swift]: this doesn't need to be under canImport

File: AdyenActions/AdyenActionComponent.swift [hunk 17]
```diff
@@ -192,7 +235,32 @@ public final class AdyenActionComponent: ActionComponent, ActionHandlingComponen
         
         currentActionComponent = weChatPaySDKActionComponent
     }
-    
+
+    #if canImport(TwintSDK)
```
Comment (erenbesel) [inline AdyenActions/AdyenActionComponent.swift]: we can put the canImport check inside the func and do nothing if sdk isn't there

File: AdyenDropIn/DropInComponent.swift [hunk 18]
```diff
@@ -234,6 +238,12 @@ public final class DropInComponent: NSObject,
             navigationController.present(asModal: component)
         case let component as InstantPaymentComponent:
             component.initiatePayment()
+        // swiftlint:disable switch_case_alignment
+        #if canImport(TwintSDK)
```
Comment (erenbesel) [inline AdyenDropIn/DropInComponent.swift]: We can remove this specific check and put TwintCompotent and InstantPaymentComponent under a protocol with `initiatePayment()`

Comment (goergisn) [reply AdyenDropIn/DropInComponent.swift]: We created a `PaymentInitiable`protocol 👌

File: AdyenActions/AdyenActionComponent.swift (hunk 17 above)
Comment (goergisn) [reply AdyenActions/AdyenActionComponent.swift]: Done 👍

File: Adyen/Assets/Generated/LocalizationKey.swift (hunk 4 above)
Comment (goergisn) [reply Adyen/Assets/Generated/LocalizationKey.swift]: That's the string that Twint provided

**Review by goergisn — DISMISSED** [review]

**Review by erenbesel — APPROVED** [review]

**Review by goergisn — APPROVED** [review]
